vm-virtio/src/console.rs

```diff
@@ -0,0 +1,353 @@
+                    }
+                }
+            }
+        }
+
+        Ok(())
+    }
+}
+
+/// Virtio device for exposing console to the guest OS through virtio.
+pub struct Console {
+    kill_evt: Option<EventFd>,
+    input_queue_evt: Option<EventFd>,
+    avail_features: u64,
+    acked_features: u64,
+    in_buffer: Arc<Mutex<VecDeque<u8>>>,
+    out: Option<Box<io::Write + Send>>,
+}
+
+impl Console {
+    /// Create a new virtio console device that gets random data from /dev/urandom.
+    pub fn new(out: Option<Box<io::Write + Send>>) -> io::Result<Console> {
+        let avail_features = 1u64 << VIRTIO_F_VERSION_1; // | (1u64 << VIRTIO_CONSOLE_F_SIZE);
+
+        Ok(Console {
+            kill_evt: None,
+            input_queue_evt: None,
+            avail_features,
+            acked_features: 0u64,
+            in_buffer: Arc::new(Mutex::new(VecDeque::new())),
+            out,
+        })
+    }
+
+    pub fn queue_bytes(&mut self, _input: &[u8]) {
+        let mut in_buffer = self.in_buffer.lock().unwrap();
+        in_buffer.extend(_input);
+        if let Some(input_queue_evt) = self.input_queue_evt.take() {
+            let _ = input_queue_evt.write(1);
+        }
+    }
+}
+
+impl Drop for Console {
+    fn drop(&mut self) {
+        if let Some(kill_evt) = self.kill_evt.take() {
+            // Ignore the result because there is nothing we can do about it.
+            let _ = kill_evt.write(1);
+        }
+    }
+}
+
+impl VirtioDevice for Console {
+    fn device_type(&self) -> u32 {
+        VirtioDeviceType::TYPE_CONSOLE as u32
+    }
+
+    fn queue_max_sizes(&self) -> &[u16] {
+        QUEUE_SIZES
+    }
+
+    fn features(&self, page: u32) -> u32 {
+        match page {
+            // Get the lower 32-bits of the features bitfield.
+            0 => self.avail_features as u32,
+            // Get the upper 32-bits of the features bitfield.
+            1 => (self.avail_features >> 32) as u32,
+            _ => {
+                warn!("Received request for unknown features page.");
+                0u32
+            }
+        }
+    }
+
+    fn ack_features(&mut self, page: u32, value: u32) {
+        let mut v = match page {
+            0 => u64::from(value),
+            1 => u64::from(value) << 32,
+            _ => {
+                warn!("Cannot acknowledge unknown features page.");
+                0u64
+            }
+        };
+
+        // Check if the guest is ACK'ing a feature that we didn't claim to have.
+        let unrequested_features = v & !self.avail_features;
+        if unrequested_features != 0 {
+            warn!("Received acknowledge request for unknown feature.");
+
+            // Don't count these features as acked.
+            v &= !unrequested_features;
+        }
+        self.acked_features |= v;
+    }
+
+    fn read_config(&self, _offset: u64, _data: &mut [u8]) {
+        warn!("No currently device specific configration defined");
```

Comment:
We should implement both `read_config()` and `write_config()` because we will support virtio-console features that requires it. (`VIRTIO_CONSOLE_F_SIZE`, `VIRTIO_CONSOLE_F_MULTIPORT `, ...)

Reply:
As per discussion with @sboeuf , we will add this feature in the next iteration.